File: 1726. Tuple with Same Product/main.py
class Solution:
    def tupleSameProduct(self, nums) -> int:

        nums.sort()
        result = 0
        n = len(nums)

        for i in range(n):
            for j in range(n - 1, i, -1):
                p = nums[i] * nums[j]

                range_set = set()
                for l in range(i + 1, j):
                    k = nums[l]

                    if p % k == 0:
                        if (p / k) in range_set:
                            result += 8
                        range_set.add(k)

        return result


if __name__ == "__main__":
    s = Solution()
    print(s.tupleSameProduct([1, 2, 4, 5, 10]))

File: 1726. Tuple with Same Product/main.go
package main

import (
	"fmt"
	"slices"
)

func tupleSameProduct(nums []int) int {

	slices.Sort(nums)
	n := len(nums)

	result := 0

	i := 0
	for i < n {
		j := i + 1
		for j < n {
			p := nums[i] * nums[j]

			range_set := make(map[int]bool)

			l := i + 1
			for l < j {

				k := nums[l]

				if p%k == 0 {
					if _, e := range_set[int(p/k)]; e {
						result += 8
					}
					range_set[k] = true
				}

				l++
			}

			j++
		}
		i++
	}

	return result
}

func main() {
	nums := []int{1, 4, 2, 5, 10}
	fmt.Println(tupleSameProduct(nums))
}
